### ISSUE TITLE
Allow to extend generated Builder

### ISSUE THREAD

**Issue by rgrigoryev** [immutables/immutables#107]
My goal is to hide Immutable related classes in my API. So I want to write something like this:
```
@Value.Immutable
@Value.Style(builder = "new")
interface My {
  class Builder extents ImmutableMy.Builder {
  }
}
```
But there are two blocks to get it working:
(1) Use only abstract Builder when hasAbstractBuilder is defined.
```ImmutableMy.Builder extends My.Builder```
This one is kind of bug and, I believe, should be fixed.

(2) Builder is hardcoded to be final
```class final lImmutableMy.Builder```
This new feature I want .

**Comment by elucash** [comment]
Hi, thank you for writing this request!
While I'm anxious about adding new style or annotation for this,
I think it should be reasonable to detect the situation
```java
interface My {
class Builder extents ImmutableMy.Builder {
// class Builder is extending ImmutableMy.Builder, so ImmutableMy.Builder will not extend Builder
// And also ImmutableMy.Builder will become non-final with package-private constructor
...
}
```

There are also some other alternatives you can consider if they applicable
See http://immutables.github.io/immutable.html#hiding-implementation and http://immutables.github.io/style.html for hide implementation or have simpler naming.
For example you can make immutable type private with standalone builder generated, where immutable object is private inside builder.

```java
@Value.Immutable
@Value.Style(visibility = ImplementationVisibility.PRIVATE)
public interface My {
}

My my = new MyBuilder().build();
```

**Comment by rgrigoryev** [comment]
Thank you for your suggestions. I've already used way described in 'Hiding implementation' section, but this leads to manually update Builder for each added/modified field. Second way unfortunately is not suitable our team codestyle.

**Comment by elucash** [comment]
This enhancement was developed and "experimentally" released in 2.0.11.
Problems we have:

1. Incomplete solution. Details are not handled well, for example, builder return types for chained invokation use immutable class (`ImmutableValue.Builder` rather than `Value.Builder`) and other things which might make usable no so pleasant.

2. Compilation breaks time after time (but not always) under Eclipse incremental compilation. Using javac in batch mode seems to work ok. After some extensive debugging it is still unclear how apt-round-processing model breaks, where excessive recompilations come from in Eclipse.

Looking forward to get the time to sort those issues out.

**Comment by rgrigoryev** [comment]
Thanks you.

**Comment by elucash** [comment]
Released in 2.0.17
